File: vllm_xpu_kernels/xpu_graph.py
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

import torch

import vllm_xpu_kernels._xpu_C  # noqa: F401


class XPUGraph:
    """Current-stream XPU graph wrapper used by vLLM."""

    def __init__(self, keep_graph: bool = False) -> None:
        self._graph = torch.classes._xpu_C.XPUGraph(keep_graph)

    def capture_begin(self, pool: tuple[int, int] | None = None) -> None:
        normalized_pool = (
            None if pool is None else (int(pool[0]), int(pool[1]))
        )
        self._graph.capture_begin(normalized_pool)

    def capture_end(self) -> None:
        self._graph.capture_end()

    def instantiate(self) -> None:
        self._graph.instantiate()

    def replay(self) -> None:
        self._graph.replay()

    def reset(self) -> None:
        self._graph.reset()

    def pool(self) -> tuple[int, int]:
        pool = self._graph.pool()
        return int(pool[0]), int(pool[1])

    def synchronize(self) -> None:
        self._graph.synchronize()

    def enable_debug_mode(self) -> None:
        self._graph.enable_debug_mode()

    def debug_dump(self, path: str) -> None:
        self._graph.debug_dump(path)


def synchronize_current_stream() -> None:
    """Wait for work submitted to the current XPU stream."""
    torch.ops._xpu_C.synchronize_current_stream()

File: csrc/xpu/xpu_graph.cpp
// SPDX-License-Identifier: Apache-2.0

#include "xpu/xpu_graph.h"

#include <c10/util/Exception.h>

VllmXpuGraph::VllmXpuGraph(bool keep_graph) : graph_(keep_graph) {}

void VllmXpuGraph::capture_begin(
    std::optional<std::tuple<std::int64_t, std::int64_t>> pool) {
  const auto capture_stream = c10::xpu::getCurrentXPUStream();
  c10::MempoolId_t pool_id{0, 0};
  if (pool.has_value()) {
    TORCH_CHECK(
        std::get<0>(*pool) >= 0 && std::get<1>(*pool) >= 0,
        "XPU graph pool values must be non-negative");
    pool_id = {
        static_cast<std::uint64_t>(std::get<0>(*pool)),
        static_cast<std::uint64_t>(std::get<1>(*pool)),
    };
  }
  graph_.capture_begin(pool_id);
  capture_stream_ = capture_stream;
}

void VllmXpuGraph::capture_end() { graph_.capture_end(); }

void VllmXpuGraph::instantiate() { graph_.instantiate(); }

void VllmXpuGraph::replay() { graph_.replay(); }

void VllmXpuGraph::reset() {
  graph_.reset();
  capture_stream_.reset();
}

std::tuple<std::int64_t, std::int64_t> VllmXpuGraph::pool() const {
  const auto pool_id = graph_.pool();
  return {
      static_cast<std::int64_t>(pool_id.first),
      static_cast<std::int64_t>(pool_id.second),
  };
}

void VllmXpuGraph::synchronize() {
  TORCH_CHECK(capture_stream_.has_value(), "XPU graph has no capture stream");
  capture_stream_->synchronize();
}

void VllmXpuGraph::enable_debug_mode() { graph_.enable_debug_mode(); }

void VllmXpuGraph::debug_dump(const std::string& path) {
  graph_.debug_dump(path);
}

void synchronize_current_stream() {
  c10::xpu::getCurrentXPUStream().synchronize();
}
